src/boot.s

```diff
@@ -10,48 +10,20 @@
 .set FPCSR_EV,  0x00000800 // Enable invalid operation exceptions
 
 // N64 PIF/OS pointers
-.set OS_MEM_SIZE,           0x80000318
 .set PIF_ENTRY_POINT,       0xBFC00000
 .set PIF_CONTROL,           0x07FC
 
 // Runtime environment pointers
 .set FS_START,              0x8000031C
 
 _start:
-    // IPL3 copied over the first 1MB. Copy over the next MB
-    // TODO: Any way to determine the actual size of the binary? We're overcopying quite a bit here
-    li $t1, 0x10101000
-    li $t3, 0x80100400
-    li $t2, 0x1FFFFF
-
-    lui $t0, 0xA460
-    sw $t1, 0x4($t0)  // cart
-    sw $t3, 0x0($t0)  // dram
-    sw $t2, 0xC($t0)  // length
-
-wait_for_dma_finished:
-    lw $t1, 0x10($t0)
-    andi $t1, $t1, 1
-    bnez $t1, wait_for_dma_finished
-    nop
+    // ipl3 writes the ramsize into 0xA4000000. Let's read it from there
+    li $a0, 0xA4000000
+    lw $a0, 0($a0)
 
     // Initialize stack
-    li $t0, OS_MEM_SIZE
-    lw $t0, 0($t0)
     li $t1, 0x80000000
-    or $sp, $t0, $t1
-
-    // Clear .bss section
-    la $t0, __bss_start
-    la $t1, __bss_end
-bss_clear_loop:
-    bge $t0, $t1, bss_clear_done
-    nop
-    sw $zero, 0($t0)
-    addiu $t0, $t0, 4
-    b bss_clear_loop
-    nop
-bss_clear_done:
+    or $sp, $t1, $a0
```

Comment:
The latest ipl3_prod also initializes $sp for you there. Actually, it also fixes a potential bug that I see here, because technically the ABI requires sp..sp+16 to be available to the callee (if it accepts arguments), and here you're not making that space available.

Reply:
Also writing 8 to PIF_ENTRY_POINT is done by IPL3

Reply:
Done